37737 fixed internal error when replacing with byte-equal parts

priera opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Bug Fix 

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed error which prevented replacing partitions when the source table had parts equal byte to byte in that partition

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

Referred in ticket #37737

priera committed:
basic implementation

alexey-milovidov commented on tests/queries/0_stateless/00626_replace_partition_from_table_zookeeper.sh:
```diff
@@ -57,6 +57,18 @@ $CLICKHOUSE_CLIENT --query="SYSTEM SYNC REPLICA dst_r2;"
 $CLICKHOUSE_CLIENT --query="SELECT count(), sum(d) FROM dst_r1;"
 $CLICKHOUSE_CLIENT --query="SELECT count(), sum(d) FROM dst_r2;"
 
+$CLICKHOUSE_CLIENT --query="SELECT 'REPLACE empty partition with duplicated parts';"
```
A new test should be added instead of modifying an existing test.

alexey-milovidov commented:
The changelog entry should explain - what exactly was fixed.

priera commented:
I'm working in the points listed above. In the meantime, here's a bit of technical description:

When running a `REPLACE PARTITION` from a source table with byte-equal parts the operation fails. The error message is like this: `Coordination::Exception. Coordination::Exception: Transaction failed: Op #3, path: /clickhouse/tables/default/mergetree1_rep/01/blocks/1_replace_from_DDB6B55418392226C72756C87A657CC1. (KEEPER_EXCEPTION)
` 

This happens because several zookeeper commands are being sent with the same `replace_from_...` path. As a solution, just sent one of such commands. The partitions are replaced in the destination table as they are in the original, i.e, no deduplication/optimization happens.

priera committed:
new test

priera committed:
removed wrong change

priera commented:
Hello, I applied both changes: moved the test to a new one and updated the changelog description. Please let me know any other change this PR may require

robot-ch-test-poll commented:
<!-- automatic status comment for PR #53735 from priera/ClickHouse:fix_zk_error_when_doing_replace -->
This is an automated comment for commit ed41b0d466ccda380a594b7dd5d8e3c8ebf4dea1 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53735/ed41b0d466ccda380a594b7dd5d8e3c8ebf4dea1/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

CurtizJ commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -7580,13 +7581,15 @@ void StorageReplicatedMergeTree::replacePartitionFrom(
                 LOG_INFO(log, "Trying to attach {} with hash_hex {}", src_part->name, hash_hex);
 
-            String block_id_path = replace ? "" : (fs::path(zookeeper_path) / "blocks" / (partition_id + "_replace_from_" + hash_hex));
+            String block_id_path = (replace || is_duplicated_part) ? "" : (fs::path(zookeeper_path) / "blocks" / (partition_id + "_replace_from_" + hash_hex));
```
But why `replace` is `false` even for `REPLACE PARTITION` query? It looks like there is a bug in another place.

priera commented:
@CurtizJ , thanks for the review and merging. I just noticed that the test I wrote won't fail in case of a regression (originally `ALTER` statement failed, but the retries mechanism used in the test hides that and the test succeeds). I will open an MR in short fixing the test
